open_spiel/python/mfg/algorithms/distribution.py

```diff
@@ -68,33 +68,34 @@ def evaluate(self):
 
     while not self.is_terminal_from_states(listing_states):
       new_listing_states = []
+      new_distribution = collections.defaultdict(float)
```

Comment:
I commented on this in the initial PR (https://github.com/deepmind/open_spiel/pull/628/commits/d058d5da4e3da6acf13dc9f110e227bb27911800):
With this new data structure for the new distribution, self.distribution no longer needs to be a default dict (which will also be safer to ensure that we don't ask for states that are not tracked, and magically get a proba of 0).

Can you also check if there are other comments I made in the initial PR that are not addressed?

Reply:
Sorry for not mentioning that. In the current code distribution is still a default dict, therefore I used also a default dict for new distribution. Also new distribution is never used to get probability (but only to update distribution) so they should not be any issue with magical 0 proba.

Reply:
Also value_str check if the key exists.

Reply:
Should I still move to a dict and not a default dict? Also should I update self.distribution to be a dict too?

Reply:
Having new_distribution be a defaultdict is good I think (otherwise, "new_distribution[new_mfg_state_str] += prob * self.value(mfg_state)" would be more complicated).

My proposal is to have "self.distribution" be a normal dict, since we no longer need the "default dict" property for that one (and we could remove the manual check in value_str()).

Reply:
Thanks for the clarification, I misunderstood your first remark. That makes sense!

Reply:
I have one question I did not manage to answer by myself: when calling the probability of each state in the distribution support, the new implementation will require that each state in the distribution support has been seen already, which might not be the case.
For example in the routing game, a possible state might be to be on link A->B. But maybe as nobody has reached the link A->B yet, the distribution has not seen A->B yet and this will raise a KeyError.
Does it make sense? Do you have any idea about potential solution (other than casting KeyError in line 88/90?